pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,321 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memberlist
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/groupcache/consistenthash"
+	"github.com/hashicorp/memberlist"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlister "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+const (
+	defaultClusterHealthzAskInterval = 40 * time.Second
+	// Set resyncPeriod to 0 to disable resyncing.
+	resyncPeriod time.Duration = 0
+	// set default virtual node replicas num of consistent hash
+	defaultVirtualNodeReplicas = 3
+)
+
+// defaultHashFn is crc32.ChecksumIEEE
+var defaultHashFn func(data []byte) uint32
+
+type ClusterInterface interface {
+	Run(stopCh <-chan struct{})
+	ShouldSelect(name string) bool
+	AddEventHandler(handler clusterNodeEventHandler)
+}
+
+type clusterNodeEventHandler func(nodeName string, isJoin bool)
+
+type Cluster struct {
+	bindPort                         int
+	NodeConfig                       *config.NodeConfig
+	nodeInformer                     coreinformers.NodeInformer
+	memberList                       *memberlist.Memberlist
+	memberRWLock                     sync.RWMutex
+	existingMembers                  []string
```

Comment:
Maybe worth to add some comments to the struct and some key variables.

Reply:
done